### ISSUE TITLE
Broken HiDPI support

### ISSUE THREAD

**Issue by jadahl** [flathub/org.telegram.desktop#305]
Seems for some reasons Telegram forces X11 on GNOME thus results in a rather terrible experience with HiDPI due to limitations of X11. I'm not aware of why this was done (some "glitches" were reported several months ago, but no description of what, and the external link to some video is broken), but using `flatpak override --user --env=QT_QPA_PLATFORM=wayland org.telegram.desktop` seems to solve all issues I'm having.

**Comment by ilya-fedin** [comment]
GNOME is blacklisted by upstream Qt
https://github.com/qt/qtbase/blob/5.15/src/gui/kernel/qguiapplication.cpp#L1402-L1427

**Comment by ilya-fedin** [comment]
Personally, I don't want to see Telegram working on Wayland natively until GNOME implement both xdg-decoration and xcursor-configuration. These two protocols are must have for Qt applications to work correctly. But it's not the full list of protocols needed for feature parity with X11, tdesktop also needs some protocol to create global shortcuts for calls, some protocol to make PiP work correctly (it respects aspect ratio on X11 while can't do the same on Wayland) and a protocol to embed windows of other Wayland connections (for webkit2gtk embed).

**Comment by jadahl** [comment]
FWIW, both on X11 and Wayland Telegram used custom window decorations, so any theoretical SSD support in GNOME wouldn't make any difference in this case. As for the rest, doesn't seem to be GNOME specific, there are no protocols finalized for any of those things. Doing some digging, the bug originally referred to for disabling Wayland on GNOME was fixed 3 years ago, the one left (something about popups) doesn't seem to be GNOME related (happens on weston) and seems to be about choosing the wrong parent surface for a popup. Lets hope things gets cleared up soon enough, Telegram via Xwayland is in my experience far worse than Wayland directly.

**Comment by ilya-fedin** [comment]
> FWIW, both on X11 and Wayland Telegram used custom window decorations, so any theoretical SSD support in GNOME wouldn't make any difference in this case.

It makes difference in the presence of "Use system window frame" option.

> As for the rest, doesn't seem to be GNOME specific, there are no protocols finalized for any of those things.

AFAIK xcursor-configuration is ready for 4 years and blocked only by GNOME

> doesn't seem to be GNOME related (happens on weston)

According to their tests, incorrect behavior happens only on GNOME, KDE seem to work just fine with their currently existing workaround.

> Lets hope things gets cleared up soon enough

I would say: don't hope. In order to fix the only remaining blocker, it looks like they need to do big internal changes in Qt and I think it's clear they won't do that just because Wayland, at least not in near time. So I would say don't expect Qt to default to Wayland on GNOME any time soon.

![изображение](https://user-images.githubusercontent.com/17829319/131488993-7b248176-dd89-4316-9758-ec6e471b4eef.png)

I think tdesktop is also affected by this bug, at least I see warnings in terminal when I open context submenus (like formatting submenu of message input field menu) on Weston

**Comment by jadahl** [comment]
> It makes difference in the presence of "Use system window frame" option.

For that the QGnomePlatform makes it work. I suspect anyone who tries to make Qt applications "blend into" GTK ones wants this anyway.

> AFAIK xcursor-configuration is ready for 4 years and blocked only by GNOME

That's not true; it's blocked on unaddressed review comments in the protocol merge request. There has been no nacks from anyone as far as I know.

> According to their tests, incorrect behavior happens only on GNOME, KDE seem to work just fine with their currently existing workaround.

The bug report seems to state that, but I could immediately reproduce the same issue with missing menus in weston, and could debug it down to being issues with incorrect toplevel used as popup parent. I know too little about Qt to know how it's possible the same issue doesn't happen on kwin.

Ok, just tried kwin, I can reproduce the same issue there. 
![Screenshot from 2021-08-31 14-03-03](https://user-images.githubusercontent.com/46741/131499210-2576d2f0-aff3-4ce3-b350-0351cd06f232.png)


> I think tdesktop is also affected by this bug, at least I see warnings in terminal when I open context submenus (like formatting submenu of message input field menu) on Weston

I see those too, but positions seems correct. Oh well, seems Qt in this case does a decent job.



....


anyway, I feel we're side tracking here :D I'll just wait for Qt to remove the old open an issue when the switch is flipped upstream.

**Comment by jadahl** [comment]
Qt5 has changed the default. Willing to backport, or will you wait for when it's part of a release?

https://codereview.qt-project.org/c/qt/qtbase/+/369443

**Comment by ilya-fedin** [comment]
it's Qt 6... And I don't think backport is a good idea, when we used KDE runtime that has a patch to make Wayland by default, people regularly reported various issues like big cursor, no 'system window frame' option and etc, loudly requesting to use XWayland by default. I don't want to step up to that again.

**Comment by jadahl** [comment]
Alright; I'll continue using the distro telegram package until it changed here then, I had to restart telegram multiple times a day due to X11 HiDPI issues. Thanks for your help.
